### PR TITLE
RFC: samples: usb: webusb: Add MS-OS descriptors for compatible IDs

### PR DESCRIPTION
**Pull request by IvanSanchez**
This adds some vendor-specific microsoft USB descriptors, which
makes the win8/win10 enumeration process to report a "WINUSB"
capability, which in turn automatically binds a WinUSB.sys driver
to a specific set of interfaces. This, in theory, makes userspace
drivers easier to handle in win32/win64 platforms.

### CODE REVIEW

**Comment by IvanSanchez** [comment]
I'll leave this as a RFC for now, as (AFAIK) we want to refactor the whole USB descriptors enchilada (#4661).

cc @finikorg @jfischer-phytec-iot @nashif @carlescufi

**Comment by carlescufi** [comment]
@IvanSanchez please add a "Signed-off-by" line in the commit message

**Comment by finikorg** [comment]
@IvanSanchez Should you handle USB String Descriptor to index 0xEE?

**Comment by IvanSanchez** [comment]
> @IvanSanchez Should you handle USB String Descriptor to index 0xEE?

@finikorg I know what you mean. Bear with me a sec.

One of the links I added in the code comments is https://docs.microsoft.com/en-us/windows-hardware/drivers/usbcon/microsoft-defined-usb-descriptors . From there, you can see that there are two versions of the MS descriptors. The first one relies on a string descriptor with a hard-coded index of `0xEE`, and the second one (implemented by this PR) relies on BOS (Binary Object Store) descriptors and vendor control requests.

So I think your question is really "Should this PR implement the version 1 MS descriptors?". To which the answer is "I don't know".

On one hand, the version 1 *should* make things work as far back as winxp (version 2 works on win8, win10). On the other hand, version 1 applies to the whole USB device, and not to specific interfaces/endpoints. And, TBH, I haven't spent too much time trying to implement/test version 1.

From my point of view, there are some questions to be tackled:

* Are the two versions of MS descriptors compatible? (*cracks knucles* - *opens text editor* - Let's find out!)

* As a firmware developer, I'd like to specify/allocate my interfaces, and let the zephyr USB stack take care of building up the descriptors, including the vendor-specific descriptors such as the MS ones. I'm not sure how to make the architecture for this.

* As this PR shows, the 1-byte codes for vendor-specific control requests are a limited resource, which some functionality might need (i.e. webusb needs one code, MS needs another code). How to make my code explicit in allocating unique values from a pool of 256?

**Comment by finikorg** [comment]
@IvanSanchez Ok, we can implement only version 2 of MS OS Descriptors (if this works fine on MS Windows 8 and 10), I was thinking to add this support for RNDIS myself.

**Comment by IvanSanchez** [comment]
@finikorg I tested this in a win10 machine, and works pretty fine. I added some "v1.0 MS OS descriptors" alongside, but haven't tested them yet.

I'm fine with having only the win8-win10 way of doing things, but unfortunately I've got the use case of supporting win7 too. I hope we can spend some time thinking on how to refactor this and make a generic solution.

(Note to self: spend a bit of time re-setting up the win7 virtual machine and wireshark)

File: samples/subsys/usb/webusb/src/main.c [hunk 1]
```diff
@@ -107,38 +107,73 @@ static const u8_t webusb_origin_url[] = {
...
+	'W',  'I',  'N',  'U',  'S',  'B',  0x00, 0x00,   /* 8-byte compatible ID string     */
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00    /* 8-byte sub-compatible ID string */
 
 };
 
 
+
+
+
 /**
```
Comment (finikorg) [inline samples/subsys/usb/webusb/src/main.c]: those empty lines are not needed

File: samples/subsys/usb/webusb/src/main.c [hunk 2]
```diff
@@ -156,6 +191,12 @@ int custom_handle_req(struct usb_setup_packet *pSetup,
 		return 0;
 	}
 
+	if (GET_DESC_TYPE(pSetup->wValue) == DESC_STRING &&
+		GET_DESC_INDEX(pSetup->wValue) == 0xEE)	{
```
[the remark this answers is not in the document]
Comment (finikorg) [reply samples/subsys/usb/webusb/src/main.c]: Though this is not always used but you should use le16_to_cpu conversion for USB packets.

**Review by finikorg — CHANGES_REQUESTED** [review]
Please split patches to logical chunks, do not include extra empty lines (never add more then one) and when not needed do not go beyond 80 characters.

**Comment by IvanSanchez** [comment]
@finikorg I've redone and force-pushed the two commits. The whitespace, extra lines, tabs and over-80-character comments should be addressed (and shippable says it's good to go).

> Though this is not always used but you should use `le16_to_cpu` conversion for USB packets.

The same is done at https://github.com/zephyrproject-rtos/zephyr/blob/3aa3e9767536ea80736bac24db73342914e24263/subsys/usb/usb_device.c#L390 , I'll leave this "as is" for now.

File: samples/subsys/usb/webusb/src/main.c [hunk 3]
```diff
@@ -24,22 +24,54 @@ static volatile bool data_transmitted;
 static volatile bool data_arrived;
 static u8_t data_buf[64];
 
-/* WebUSB Platform Capability Descriptor */
+/* WebUSB + MS Platform Capability Descriptors */
 static const u8_t webusb_bos_descriptor[] = {
-	/* Binary Object Store descriptor */
-	0x05, 0x0F, 0x1D, 0x00, 0x01,
+
```
Comment (finikorg) [inline samples/subsys/usb/webusb/src/main.c]: why do you need here empty line?

Comment (finikorg) [reply samples/subsys/usb/webusb/src/main.c]: Moreover this useful change does not belong to this particular patch.

File: samples/subsys/usb/webusb/src/main.c [hunk 4]
```diff
@@ -122,6 +181,12 @@ int vendor_handle_req(struct usb_setup_packet *pSetup,
 		*data = (u8_t *)(&webusb_origin_url);
 		*len = sizeof(webusb_origin_url);
 
+		return 0;
```
Comment (finikorg) [inline samples/subsys/usb/webusb/src/main.c]: here there is an empty line before return

File: samples/subsys/usb/webusb/src/main.c [hunk 5]
```diff
@@ -122,6 +181,12 @@ int vendor_handle_req(struct usb_setup_packet *pSetup,
...
+	} else if (pSetup->bRequest == 0x02 && pSetup->wIndex == 0x07) {
+		/* Get MS OS 2.0 Descriptors request */
+		/* 0x07 means "MS_OS_20_DESCRIPTOR_INDEX" */
+		*data = (u8_t *)(&msos2_descriptor);
+		*len = sizeof(msos2_descriptor);
 		return 0;
```
Comment (finikorg) [inline samples/subsys/usb/webusb/src/main.c]: return with no empty line before. different style in small chunk of the code!

**Review by finikorg — COMMENTED** [review]
Second patch looks good now ;)

File: samples/subsys/usb/webusb/src/main.c (hunk 3 above)
Comment (IvanSanchez) [reply samples/subsys/usb/webusb/src/main.c]: OK, I'll split this.

File: samples/subsys/usb/webusb/src/main.c [hunk 6]
```diff
@@ -24,7 +24,7 @@ static volatile bool data_transmitted;
 static volatile bool data_arrived;
 static u8_t data_buf[64];
 
-/* WebUSB Platform Capability Descriptor */
+/* WebUSB + MS Platform Capability Descriptors */
 static const u8_t webusb_bos_descriptor[] = {
 	/* Binary Object Store descriptor */
 	0x05, 0x0F, 0x1D, 0x00, 0x01,
```
Comment (IvanSanchez) [inline samples/subsys/usb/webusb/src/main.c]: Wait wait wait, by reverting the cosmetic changes I also reverted the length and number of following capability descriptors.

File: samples/subsys/usb/webusb/src/main.c [hunk 7]
```diff
@@ -24,10 +24,10 @@ static volatile bool data_transmitted;
 static volatile bool data_arrived;
 static u8_t data_buf[64];
 
-/* WebUSB Platform Capability Descriptor */
+/* WebUSB + MS Platform Capability Descriptors */
 static const u8_t webusb_bos_descriptor[] = {
 	/* Binary Object Store descriptor */
-	0x05, 0x0F, 0x1D, 0x00, 0x01,
+	0x05, 0x0F, 0x39, 0x00, 0x02,
```
Comment (IvanSanchez) [inline samples/subsys/usb/webusb/src/main.c]: This *will* create a merge conflict with #7094. :cry:

Comment (finikorg) [reply samples/subsys/usb/webusb/src/main.c]: I was asking to make a separate patch, you can still merge those two branches

Comment (finikorg) [reply samples/subsys/usb/webusb/src/main.c]: If there is a PR already then just move that path to be the first in this series.

Comment (IvanSanchez) [reply samples/subsys/usb/webusb/src/main.c]: OK, this calls for a bit of clever rebasing.

**Comment by carlescufi** [comment]
Please update the copyright in the file with a line including Nordic Semiconductor

**Review by carlescufi — CHANGES_REQUESTED** [review]
Change email to be the Nordic Semi one in your Signed-off line

**Comment by carlescufi** [comment]
@IvanSanchez will you address the comments raised so far?

**Review by finikorg — APPROVED** [review]

**Comment by finikorg** [comment]
This PR can be closed now, the patches were applied through other PRs.

**Pull request closed without merging**